### PR TITLE
[mono] Add RunCoreFxTest AOT targets to mono.proj (+ more rules for Makefile)

### PR DESCRIPTION
**Pull request by EgorBo**
Add more targets and rules
1) Now it's possible to build & run a specific corefx test suite via makefile, e.g. a typical worflow may look like:
```bash
# change something in corelib
make corelib
# change something in System.Linq.Test
make run-test-corefx LibraryToTest=System.Linq
```
2) Add AOT target for `.dotnet-mono` (to precompile all BCL libs inside `.dotnet-mono`)
and AOT target for corefx tests (to precompile all CoreFX tests)

### CODE REVIEW

File: src/mono/mono.proj [hunk 1]
```diff
@@ -111,17 +114,74 @@
...
+  <Target Name="PrecompileAssembly">
+   <PropertyGroup>
+      <EnvVars Condition="'$(OS)' != 'Windows_NT'">
+      <![CDATA[MONO_ENV_OPTIONS="--aot=mcpu=native"]]>
+      </EnvVars>
+      <EnvVars Condition="'$(OS)' == 'Windows_NT'">
+      <![CDATA[set "MONO_ENV_OPTIONS=--aot=mcpu=native"]]>
+      </EnvVars>
+    </PropertyGroup>
+    <Exec Command="$(EnvVars) $(MonoExec) %(AssembliesToPrecompile.Identity)" ContinueOnError="WarnAndContinue" />
```
Comment (marek-safar) [inline src/mono/mono.proj]: I'd be really great if we could avoid doing it this way and use proper publish step

Comment (EgorBo) [reply src/mono/mono.proj]: I agree but it's at the moment it's the only way, but already looking for better approaches.

File: src/mono/mono.proj [hunk 2]
```diff
@@ -151,10 +196,16 @@
 
   <!-- Copy Mono runtime bits to the coreclr's Core_Root in order to run runtime tests -->
   <Target Name="PatchCoreClrCoreRoot" DependsOnTargets="ValidateLocalDotnet">
-    <PropertyGroup>
-      <Destination>$(CoreClrTestCoreRoot)</Destination>
-    </PropertyGroup>
-    <CallTarget Targets="CopyMonoRuntimeFilesFromArtifactsToDestination" />
+    <MSBuild Projects ="$(MSBuildProjectFullPath)"
```
Comment (akoeplinger) [inline src/mono/mono.proj]:
```suggestion
    <MSBuild Projects="$(MSBuildProjectFullPath)"
```

File: src/mono/mono.proj [hunk 3]
```diff
@@ -151,10 +196,16 @@
...
+            Properties="Destination=$(CoreClrTestCoreRoot)"
+            Targets="CopyMonoRuntimeFilesFromArtifactsToDestination" />
+  </Target>
+
+  <!-- Copy Mono runtime bits to the coreclr's Core_Root in order to run runtime tests -->
+  <Target Name="PatchCoreFxTesthost" DependsOnTargets="ValidateLocalDotnet">
+    <MSBuild Projects ="$(MSBuildProjectFullPath)"
```
Comment (akoeplinger) [inline src/mono/mono.proj]:
```suggestion
    <MSBuild Projects="$(MSBuildProjectFullPath)"
```

File: src/mono/mono.proj [hunk 4]
```diff
@@ -190,6 +241,14 @@
     <Exec Condition="'$(OS)' != 'Windows_NT'" Command="$(MonoProjectRoot)../coreclr/tests/./runtest.sh $(CoreClrTestConfig)" ContinueOnError="ErrorAndContinue" />
   </Target>
 
+  <!-- Run corefx tests using testhost -->
+  <Target Name="RunCoreFxTest" DependsOnTargets="PatchCoreFxTesthost">
+    <Error Condition="$(LibraryToTest) == ''" Text="LibraryToTest variable is not set" />
+    <Error Condition="$(LibraryToTest.EndsWith('.Tests'))" Text="LibraryToTest should not end with .Tests, e.g. just 'System.Runtime'" />
+    <MSBuild Projects ="$(RepoRoot)src\libraries\$(LibraryToTest)\tests\$(LibraryToTest).Tests.csproj"
```
Comment (akoeplinger) [inline src/mono/mono.proj]:
```suggestion
    <MSBuild Projects ="$(RepoRoot)src\libraries\$(LibraryToTest)\tests\$(LibraryToTest).Tests.csproj" 
```

File: src/mono/netcore/Makefile [hunk 5]
```diff
@@ -14,28 +15,39 @@ ifeq ($(words $(wildcard ../../../artifacts/tests/coreclr/*.*.*)), 1)
...
+endif
+
 MONO_PROJ=/p:CoreClrTestConfig=$(CORECLR_TESTS_CONFIG) /p:Configuration=$(MONO_RUNTIME_CONFIG) ../mono.proj
 
 # run sample using local .dotnet-mono
 # build it with .dotnet first in order to be able to use LLVM only for the actual sample
 run-sample: patch-mono-dotnet
 	$(DOTNET) build -c Release sample/HelloWorld
-	MONO_ENV_OPTIONS="" COMPlus_DebugWriteToStdErr=1 \
+	MONO_ENV_OPTIONS="--aot" COMPlus_DebugWriteToStdErr=1 \
```
Comment (akoeplinger) [inline src/mono/netcore/Makefile]: why do we want to default to --aot here?

**Review by akoeplinger — CHANGES_REQUESTED** [review]

File: src/mono/netcore/Makefile [hunk 6]
```diff
@@ -49,6 +61,15 @@ run-tests-coreclr:
 run-tests-coreclr-all:
 	$(DOTNET) msbuild /t:RunCoreClrTests $(MONO_PROJ)
 
+# run corefx tests for $(LibraryToTest)
+# e.g. `make run-tests-corefx LibraryToTest=System.Runtime`
+run-tests-corefx:
+	$(DOTNET) msbuild /t:RunCoreFxTest /p:LibraryToTest=$(LibraryToTest) $(MONO_PROJ)
+
```
Comment (vargaz) [inline src/mono/netcore/Makefile]: The make variable name should be uppercase and shorter.

File: src/mono/netcore/Makefile (hunk 5 above)
Comment (EgorBo) [reply src/mono/netcore/Makefile]: It's supposed to be `--llvm` (just to showcase how to use mono cli). If mono runtime is built without llvm it won't crash, will just work normally + print "runtime doesn't support llvm"

File: src/mono/netcore/Makefile [hunk 7]
```diff
@@ -49,6 +61,16 @@ run-tests-coreclr:
 run-tests-coreclr-all:
 	$(DOTNET) msbuild /t:RunCoreClrTests $(MONO_PROJ)
 
+# run corefx tests for a bcl lib, e.g.:
+#   make run-tests-corefx-System.Runtime 
+# for `System.Runtime.Tests`
+run-tests-corefx-%:
+	$(DOTNET) msbuild /t:RunCoreFxTest /p:LibraryToTest=$* $(MONO_PROJ)
```
Comment (akoeplinger) [inline src/mono/netcore/Makefile]:
```suggestion
	$(DOTNET) msbuild /t:RunLibrariesTest /p:LibraryToTest=$* $(MONO_PROJ)
```

Comment (EgorBo) [reply src/mono/netcore/Makefile]: yeah I am still cleaning it all up :-)

File: src/mono/mono.proj [hunk 8]
```diff
@@ -119,6 +121,49 @@
...
+  <Target Name="PrecompileLocalMonoDotnetBcl" DependsOnTargets="PrepareLocalMonoDotnetBclForAot">
+    <CallTarget Targets="PrecompileAssembly"/>
+  </Target>
+  <Target Name="PrepareLocalMonoDotnetBclForAot" DependsOnTargets="PatchLocalMonoDotnet">
+    <PropertyGroup>
+      <MonoExec>$(LocalMonoDotnet)</MonoExec>
+    </PropertyGroup>
+    <ItemGroup>
+      <AssembliesToPrecompile 
+          Include="$(LocalMonoDotnetDir)\shared\Microsoft.NETCore.App\**\*.dll"
```
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/mono/mono.proj]: Will do a bit later, it's an initial valid implementation for now 🙂

**Review by akoeplinger — APPROVED** [review]

**Pull request merged**